pkg/apis/metrics/types.go

```diff
@@ -16,14 +16,427 @@ limitations under the License.
+	// Number of connections in the TimeWait state.
+	TimeWait uint64 `json:"timeWait,omitempty"`
+	// Number of connections in the Close state.
+	Close uint64 `json:"close,omitempty"`
+	// Number of connections in the CloseWait state.
+	CloseWait uint64 `json:"closeWait,omitempty"`
+	// Number of connections in the LastAck state.
+	LastAck uint64 `json:"lastAck,omitempty"`
+	// Number of connections in the Listen state.
+	Listen uint64 `json:"listen,omitempty"`
+	// Number of connections in the Closing state.
+	Closing uint64 `json:"closing,omitempty"`
+}
+
+type NetworkStats struct {
+	// Network stats by interface.
+	Interfaces []InterfaceStats `json:"interfaces,omitempty"`
+	// TCP connection stats (Established, Listen...)
+	Tcp TcpStat `json:"tcp,omitempty"`
+	// TCP6 connection stats (Established, Listen...)
+	Tcp6 TcpStat `json:"tcp6,omitempty"`
+}
+
+type InterfaceStats struct {
+	// The name of the interface.
+	Name string `json:"name,omitempty"`
+	// Cumulative count of bytes received.
+	RxBytes uint64 `json:"rxBytes,omitempty"`
+	// Cumulative count of packets received.
+	RxPackets uint64 `json:"rxPackets,omitempty"`
+	// Cumulative count of receive errors encountered.
+	RxErrors uint64 `json:"rxErrors,omitempty"`
+	// Cumulative count of packets dropped while receiving.
+	RxDropped uint64 `json:"rxDropped,omitempty"`
+	// Cumulative count of bytes transmitted.
+	TxBytes uint64 `json:"txBytes,omitempty"`
+	// Cumulative count of packets transmitted.
+	TxPackets uint64 `json:"txPackets,omitempty"`
+	// Cumulative count of transmit errors encountered.
+	TxErrors uint64 `json:"txErrors,omitempty"`
+	// Cumulative count of packets dropped while transmitting.
+	TxDropped uint64 `json:"txDropped,omitempty"`
+}
+
+// Instantaneous CPU stats
+type CpuInstStats struct {
+	Usage CpuInstUsage `json:"usage,omitempty"`
+}
+
+// CPU usage time statistics.
+type CpuInstUsage struct {
+	// Total CPU usage.
+	// Units: nanocores per second
+	Total uint64 `json:"total,omitempty"`
+
+	// Per CPU/core usage of the container.
+	// Unit: nanocores per second
+	PerCpu []uint64 `json:"perCpu,omitempty"`
+
+	// Time spent in user space.
+	// Unit: nanocores per second
+	User uint64 `json:"user,omitempty"`
+
+	// Time spent in kernel space.
+	// Unit: nanocores per second
+	System uint64 `json:"system,omitempty"`
+}
+
+// CPU usage time statistics.
+type CpuUsage struct {
+	// Total CPU usage.
+	// Units: nanoseconds
+	Total uint64 `json:"total,omitempty"`
+
+	// Per CPU/core usage of the container.
+	// Unit: nanoseconds.
+	PerCpu []uint64 `json:"perCpu,omitempty"`
+
+	// Time spent in user space.
+	// Unit: nanoseconds
+	User uint64 `json:"user,omitempty"`
+
+	// Time spent in kernel space.
+	// Unit: nanoseconds
+	System uint64 `json:"system,omitempty"`
+}
+
+// All CPU usage metrics are cumulative from the creation of the container
+type CpuStats struct {
+	Usage CpuUsage `json:"usage,omitempty"`
+	// Smoothed average of number of runnable threads x 1000.
```

Comment:
`Core` can mean different things. Whereas `Cpu` is unambiguous.

On Thu, Nov 12, 2015 at 3:50 PM, Tim St. Clair <notifications@github.com>
wrote:

> In pkg/apis/metrics/types.go
> <https://github.com/kubernetes/kubernetes/pull/16458#discussion_r44733131>
> :
>
> > +	// Unit: nanoseconds.
> > +	PerCpu []uint64 `json:"perCpu,omitempty"`
> > +
> > +	// Time spent in user space.
> > +	// Unit: nanoseconds
> > +	User uint64 `json:"user,omitempty"`
> > +
> > +	// Time spent in kernel space.
> > +	// Unit: nanoseconds
> > +	System uint64 `json:"system,omitempty"`
> > +}
> > +
> > +// All CPU usage metrics are cumulative from the creation of the container
> > +type CpuStats struct {
> > +	Usage CpuUsage `json:"usage,omitempty"`
> > +	// Smoothed average of number of runnable threads x 1000.
>
> Maybe the comment can be updated to reflect that. Can we either ban the
> word "CPU" in this file (use "core" instead), or explicitly define it?
>
> I think CPU should refer to the abstract resource (e.g. CPU stats), while
> core refers to the actual measurement (core is a unit). I updated a couple
> of comments to reflect that. Does it make sense?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16458/files#r44733131>.
>